lib/streamaggr/rate.go

```diff
@@ -1,183 +1,192 @@
-			sv.lastValues[inputKey] = lv
-			sv.deleteDeadline = deleteDeadline
-		}
-		sv.mu.Unlock()
-		if deleted {
-			// The entry has been deleted by the concurrent call to flushState
-			// Try obtaining and updating the entry again.
-			goto again
+func (v *rateAggrSharedValue) getState(isGreen bool) *rateAggrStateValue {
+	if isGreen {
+		if v.green == nil {
+			v.green = getRateAggrStateValue()
 		}
+		return v.green
 	}
+	if v.blue == nil {
+		v.blue = getRateAggrStateValue()
+	}
+	return v.blue
 }
 
-func (as *rateAggrState) flushState(ctx *flushCtx) {
-	currentTime := fasttime.UnixTimestamp()
-
-	suffix := as.getSuffix()
+func (v *rateAggrSharedValue) reset() {
+	v.value = 0
+	v.deleteDeadline = 0
+	v.prevTimestamp = 0
+	putRateAggrStateValue(v.blue)
+	putRateAggrStateValue(v.green)
+	v.blue = nil
+	v.green = nil
+}
 
-	as.removeOldEntries(currentTime)
+type rateAggrStateValue struct {
+	// increase stores cumulative increase for the current time series on the current aggregation interval
+	increase  float64
+	timestamp int64
+}
 
-	m := &as.m
-	m.Range(func(k, v any) bool {
-		sv := v.(*rateStateValue)
+type rateAggrValue struct {
+	shared  map[string]*rateAggrSharedValue
+	isGreen bool
+}
 
-		sv.mu.Lock()
-		lvs := sv.lastValues
-		sumRate := 0.0
-		countSeries := 0
-		for k1, lv := range lvs {
-			d := float64(lv.timestamp-lv.prevTimestamp) / 1000
-			if d > 0 {
-				sumRate += lv.increase / d
-				countSeries++
-			}
-			lv.prevTimestamp = lv.timestamp
-			lv.increase = 0
-			lvs[k1] = lv
+func (av *rateAggrValue) pushSample(_ aggrConfig, sample *pushSample, key string, deleteDeadline int64) {
+	var state *rateAggrStateValue
+	sv, ok := av.shared[key]
+	if ok {
+		state = sv.getState(av.isGreen)
+		if sample.timestamp < state.timestamp {
+			// Skip out of order sample
+			return
 		}
-		deleted := sv.deleted
-		sv.mu.Unlock()
-
-		if countSeries == 0 || deleted {
-			// Nothing to update
-			return true
+		if sample.value >= sv.value {
+			state.increase += sample.value - sv.value
+		} else {
+			// counter reset
+			state.increase += sample.value
 		}
+	} else {
+		sv = getRateAggrSharedValue(av.isGreen)
+		sv.prevTimestamp = sample.timestamp
+		key = bytesutil.InternString(key)
+		av.shared[key] = sv
+		state = sv.getState(av.isGreen)
+	}
+	sv.value = sample.value
+	sv.deleteDeadline = deleteDeadline
+	state.timestamp = sample.timestamp
+}
 
-		result := sumRate
-		if as.isAvg {
-			result /= float64(countSeries)
+func (av *rateAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+	ac := c.(*rateAggrConfig)
+	var state *rateAggrStateValue
+	suffix := ac.getSuffix()
+	rate := 0.0
+	countSeries := 0
+	for sk, sv := range av.shared {
+		if ctx.flushTimestamp > sv.deleteDeadline {
+			delete(av.shared, sk)
+			putRateAggrSharedValue(sv)
+			continue
+		}
+		if sv.prevTimestamp == 0 {
+			continue
+		}
+		if av.isGreen {
+			state = sv.green
+		} else {
+			state = sv.blue
+		}
+		if state == nil {
```

Comment:
In which case state could be nil?

Reply:
removed, leftover from previous attempts